src/NUnitFramework/tests/AsyncExecutionApiAdapter.cs

```diff
@@ -0,0 +1,176 @@
+// ***********************************************************************
+// Copyright (c) 2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+#if ASYNC
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using NUnit.Framework.Constraints;
+using NUnit.Framework.Internal;
+using NUnit.TestUtilities;
+
+namespace NUnit.Framework
+{
+    /// <summary>
+    /// Adapts each NUnit API that executes async user code synchronously to a common interface
+    /// to be tested for behavior.
+    /// </summary>
+    public abstract class AsyncExecutionApiAdapter
+    {
+        public static IEnumerable<AsyncExecutionApiAdapter> All { get; } = new AsyncExecutionApiAdapter[]
+        {
+            new TaskReturningTestMethodAdapter(),
+            new AssertThrowsAsyncAdapter(),
+            new AssertDoesNotThrowAsyncAdapter(),
+            new AssertCatchAsyncAdapter(),
+            new ThrowsNothingConstraintAdapter(),
+            new ThrowsConstraintAdapter(),
+            new ThrowsExceptionConstraintAdapter(),
+            new NormalConstraintAdapter()
+        };
```

Comment:
I like these! 😄

Reply:
Oops, I just realized I forgot to cover setups and teardowns!

Reply:
Yeah, isn't this better than what I tried to do the last time? 😆
Oops, I just realized I forgot to cover setups and teardowns!